File: sdk/dotnet/src/AgentHooks/Canonical.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
// Canonical JSON serialization and context identity (§10).
//
// Delegates to the Rust core via libagent_hooks_ffi so every SDK
// produces byte-identical output. The pure-C# implementation was
// removed once the core became canonical (see
// sdk/rust/core/src/canonical.rs).

using System.Text.Json;
using System.Text.Json.Nodes;

namespace AgentHooks;

public static class Canonical
{
    private static readonly JsonSerializerOptions Compact = new() { WriteIndented = false };

    /// <summary>Serialize per §10.1. Implemented by the Rust core.</summary>
    public static string Json(JsonNode? node) =>
        Native.CanonicalJson(node?.ToJsonString(Compact) ?? "null");

    /// <summary><c>"sha256:" + hex(SHA-256(Json(ctx_L01)))</c> (§10.2). Rust core.</summary>
    public static string ContextIdentity(AgentContext ctx) =>
        Native.ContextIdentity(ctx.Json.ToJsonString(Compact));

    /// <summary>§5: validate an interceptor's wire return. Rust core.</summary>
    public static void ValidateVerdict(JsonNode verdict) =>
        Native.ValidateVerdict(verdict.ToJsonString(Compact));

    /// <summary>§5.2: apply a <c>$target</c>-rooted transform. Rust core.</summary>
    public static JsonNode? ApplyTransform(JsonNode? target, string path, JsonNode? value) =>
        JsonNode.Parse(Native.ApplyTransform(
            target?.ToJsonString(Compact) ?? "null",
            path,
            value?.ToJsonString(Compact) ?? "null"));

    /// <summary>§7.1 fold-through: apply one transform to the context's
    /// target (and its L1 alias). Returns the updated context. Rust core.</summary>
    public static JsonObject ApplyTransformCtx(AgentContext ctx, string path, JsonNode? value) =>
        (JsonObject)JsonNode.Parse(Native.ApplyTransformCtx(
            ctx.Json.ToJsonString(Compact),
            path,
            value?.ToJsonString(Compact) ?? "null"))!;

    /// <summary>§8 <c>evaluate_only</c>: validate a transform against the
    /// context's target without applying it. Rust core.</summary>
    public static void ValidateTransformCtx(AgentContext ctx, string path, JsonNode? value) =>
        Native.ValidateTransformCtx(
            ctx.Json.ToJsonString(Compact),
            path,
            value?.ToJsonString(Compact) ?? "null");

    /// <summary>§6/§10.3: build the <c>InterceptionRecord</c> for one completed
    /// emission. <paramref name="options"/> is the ah_finalize options object
    /// (<c>input_identity</c>, <c>identity_provider</c>, <c>enforced_identity</c>,
    /// <c>decided_by</c>, <c>composition</c> (REQUIRED), <c>verdicts</c>,
    /// <c>fold_truncated</c>, <c>resolved_by</c>). Rust core.</summary>
    public static JsonObject Finalize(
        AgentContext ctx, JsonNode verdict, EnforcementMode mode, JsonObject options) =>
        (JsonObject)JsonNode.Parse(Native.Finalize(
            ctx.Json.ToJsonString(Compact),
            verdict.ToJsonString(Compact),
            mode == EnforcementMode.Enforce ? "enforce" : "evaluate_only",
            options.ToJsonString(Compact)))!;

    /// <summary>§7.3/§7.5: severity-max aggregation for the multi-verdict
    /// composition profiles. Returns <c>{combined, decided_by, consult,
    /// apply_transform, verdicts}</c>. Rust core.</summary>
    public static JsonObject ComposeAggregate(CompositionConfig composition, JsonArray verdicts) =>
        (JsonObject)JsonNode.Parse(Native.ComposeAggregate(
            composition.ToWire().ToJsonString(Compact),
            verdicts.ToJsonString(Compact)))!;
}

File: sdk/go/agenthooks/canonical.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package agenthooks

// Canonical JSON serialization and context identity (§10).
//
// Delegates to the Rust core via cgo (libagent_hooks_ffi) so every SDK
// produces byte-identical output. The pure-Go implementation was removed
// once the core became canonical (see sdk/rust/core/src/canonical.rs).

import "encoding/json"

// CanonicalJSON serializes v per §10.1. Implemented by the Rust core.
func CanonicalJSON(v any) (string, error) {
	b, err := json.Marshal(v)
	if err != nil {
		return "", err
	}
	return nativeCanonicalJSON(string(b))
}

// ContextIdentity returns "sha256:" + hex(SHA-256(CanonicalJSON(ctx_L01)))
// (§10.2). Implemented by the Rust core.
func ContextIdentity(ctx AgentContext) (string, error) {
	b, err := json.Marshal(map[string]any(ctx))
	if err != nil {
		return "", err
	}
	return nativeContextIdentity(string(b))
}

// ApplyTransform applies a $target-rooted path to target and returns the
// new target (§5.2). Implemented by the Rust core.
func ApplyTransform(target any, path string, value any) (any, error) {
	tb, err := json.Marshal(target)
	if err != nil {
		return nil, err
	}
	vb, err := json.Marshal(value)
	if err != nil {
		return nil, err
	}
	out, err := nativeApplyTransform(string(tb), path, string(vb))
	if err != nil {
		return nil, err
	}
	var result any
	return result, json.Unmarshal([]byte(out), &result)
}

// ValidateVerdict validates an interceptor's wire return per §5.
// Implemented by the Rust core.
func ValidateVerdict(v Verdict) error {
	b, err := json.Marshal(v)
	if err != nil {
		return err
	}
	_, err = nativeValidateVerdict(string(b))
	return err
}

// CombineVerdicts combines an ordered slice of verdicts per §7.1.
// Implemented by the Rust core.
func CombineVerdicts(vs []Verdict) (Verdict, error) {
	b, err := json.Marshal(vs)
	if err != nil {
		return Verdict{}, err
	}
	out, err := nativeCombineVerdicts(string(b))
	if err != nil {
		return Verdict{}, err
	}
	var v Verdict
	return v, json.Unmarshal([]byte(out), &v)
}
